clients/unshaded/src/main/java/tachyon/client/block/TachyonBlockStore.java

```diff
@@ -89,25 +90,29 @@ public BufferedBlockInStream getInStream(long blockId) throws IOException {
     try {
       // TODO(calvin): Fix this RPC.
       BlockInfo blockInfo = masterClient.getBlockInfo(blockId);
-      // TODO(calvin): Get location via a policy.
       if (blockInfo.locations.isEmpty()) {
-        // TODO(calvin): Maybe this shouldn't be an exception.
-        throw new IOException("No block " + blockId + " is not available in Tachyon");
+        throw new IOException("Block " + blockId + " is not available in Tachyon");
       }
-      // TODO(calvin): Investigate making this a Factory method
-      NetAddress workerNetAddress = blockInfo.locations.get(0).getWorkerAddress();
-      InetSocketAddress workerAddr =
-          new InetSocketAddress(workerNetAddress.getHost(), workerNetAddress.getDataPort());
-      if (NetworkAddressUtils.getLocalHostName(ClientContext.getConf()).equals(
-          workerAddr.getHostName())) {
-        if (mContext.hasLocalWorker()) {
-          return new LocalBlockInStream(blockId, blockInfo.getLength(), workerAddr);
-        } else {
-          throw new IOException("Local read requested but there is no local worker.");
+      // TODO(calvin): Get location via a policy.
+      // Although blockInfo.locations are sorted by tier, we prefer reading from the local worker.
+      // But when there is no local worker or there are no local blocks, we prefer the first
+      // location in blockInfo.locations that is nearest to memory tier.
+      // Assuming if there is no local worker, there are no local blocks in blockInfo.locations.
+      if (mContext.hasLocalWorker()) {
```

Comment:
`mContext.hasLocalWorker()` is actually not a great method at the moment since its return value is initialized when the client is started and could be stale. Searching through the block locations is pretty cheap, so it's probably makes sense to avoid checking mContext.hasLocalWorker() at all.